Hapi v17

rafaelsouzaf opened the pull request:
(UNDER DEVELOPMENT)

- [x ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/hapijs/hapi/blob/master/API.md
- [ x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

BorntraegerMarc commented on types/hapi/v17/hapi-tests.ts:
I would import it here with `import { Server } from 'hapi'` Then instanciate it with:
```
import { Server } from 'hapi';
const server = new Server({});
```

BorntraegerMarc commented on types/hapi/v17/index.d.ts:
Aren't the other exports missing?

rafaelsouzaf commented on types/hapi/v17/hapi-tests.ts:
Oh yes. I didn't work with this file yet. It's your original code from "hapi-17.0-marc" branch. I can remove this file from the PR if you want. What do you think?

rafaelsouzaf commented on types/hapi/v17/index.d.ts:
Of course, is almost the same here. I didn't work with this file yet. I only added your profile in "Definitions by" I guess.

rafaelsouzaf commented:
@BorntraegerMarc 

> Did you decide now to do TS 2.6 with Hapi v16 first and then move to Hapi v17?

Oh no. Never. I am working with v17 and TS 2.6 only.

> Because in this PR there are also a lot of Hapi v16 changes.

It's because I am following the recommendation of @AJamesPhillips in the issue:
(https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20670#issuecomment-337924152)

To start working with the v17, first...

> "... we'd make a v16 folder first and copy the current index.d.ts and tests to that, then update the root level index.d.ts, tests and tsconfig to work with typescript 2.6 and hapi v17."

So I created the v16 folder, moved v16 files to inside and changed the paths to keep working fine :-)
I think we talked about that before. Maybe I am not understanding very well the recommendation.

> I would maybe suggest that we jump directly to v17 and leave v16 be (because it seems to work well for now...) Sorry for changing my opinion again :)

No problem. I liked :)

> But definitely the v16 changes need to be deleted from this PR. If we want to do them it needs to be in a separate PR

Ok no problem. Just to confirm: So, I need to keep in this PR only the v17 folder, right?

BorntraegerMarc commented:
@rafaelsouzaf 

>  So, I need to keep in this PR only the v17 folder, right?
Yes 😄

rafaelsouzaf commented:
Thanks @BorntraegerMarc I appreciate if you can review the branch now. Please let me know if you have any question.

BorntraegerMarc commented on types/hapi/v17/definitions/plugin/plugin.d.ts:
is this really a todo?

BorntraegerMarc commented on types/hapi/v17/definitions/request/request.d.ts:
You could define the type as `any` in worst case....
Should we post it in https://github.com/hapijs/hapi/issues/3669 ?
And did you also not find it in the tests?

BorntraegerMarc commented on types/hapi/v17/definitions/request/request.d.ts:
Could you define it possibly as `() => Object`?

BorntraegerMarc commented on types/hapi/v17/definitions/route/route-options-payload.d.ts:
Don't you think it should look a  bit more like:
```
export enum PayloadMultipartOutputEnum {
  output {
    Data = 'data',
    Stream = 'stream',
    File = 'file',
    Annotated : {
      headers: '',
      filename: '',
      payload: ''
    }
  }
}
```

however the `Annotated` options can be whatever the function caller chooses. So maybe an interface is better fit....

SimonSchick commented on types/hapi/v17/definitions/server/server-connection-options.d.ts:
What's with the inconsistent interface names?

rafaelsouzaf commented on types/hapi/v17/definitions/server/server-connection-options.d.ts:
@SimonSchick The inconsistent names between for example RouteOptions (starting without "I") and IServerConnectionOptions (starting with "I") right? Yes, we are working on that :-)

SimonSchick commented on types/hapi/v17/definitions/server/server-connection-options.d.ts:
Hah that's fine, I read this diff late at night and thought this already went into DT main repo 😛

rafaelsouzaf commented on types/hapi/v17/definitions/plugin/plugin.d.ts:
Fixed. Please see the new PR (https://github.com/rafaelsouzaf/DefinitelyTyped/pull/2/)
Sorry. I should not have closed this one.

rafaelsouzaf commented on types/hapi/v17/definitions/request/request.d.ts:
Yes, maybe we can post there. But I would like to advance a little more. Some parts of the code and API are more "clear" when I review it. I will keep the TODO for now.

rafaelsouzaf commented on types/hapi/v17/definitions/request/request.d.ts:
Can you give me an example please?

rafaelsouzaf commented on types/hapi/v17/definitions/route/route-options-payload.d.ts:
I don't know if we can use Enum with this format. Are you sure?

With Java I know very well when we should use Enum or not. With TS I am a little bit confused yet because I think some times we can use Interface or Type with the same objective. I am using a lot of Enums in the definitions, but if you like we can review it and maybe refactoring everything.

rafaelsouzaf commented on types/hapi/v17/definitions/server/server-connection-options.d.ts:
@SimonSchick when you want, please check the PR (https://github.com/rafaelsouzaf/DefinitelyTyped/pull/2/)

rafaelsouzaf committed:
HapiJs v17 Typescript Definitions. Recovering the repo.

rafaelsouzaf committed:
More Server Definitions.

rafaelsouzaf committed:
Added more Server and Plugins definitions.

rafaelsouzaf committed:
Updated imports with new definitions

rafaelsouzaf committed:
Added ResponseToolkit def.

Commit:
Added some events in ServerEvents interface. Not ready yet.

Commit:
Removing unused param from JSDoc.

rafaelsouzaf committed:
Added more Server definitions. Moved the events from ServerEvents.d.ts to Server.d.ts.

rafaelsouzaf committed:
Removed legacy definitions.

rafaelsouzaf committed:
Replaced enum HttpMethod to types like v16. Not very elegant solution but works easily with lower/upper case :)

rafaelsouzaf committed:
Separating request auth, server and route in different files to better understanding.

rafaelsouzaf committed:
Added more Server definitions: ServerExt, ServerInject and others.

rafaelsouzaf committed:
Replaced "any" type to "RouteCompressionEncoderSettings" type to better understanding because is shared with other interface.

rafaelsouzaf committed:
Updated index with last file definitions.

rafaelsouzaf committed:
Added generic LifeCycle interfaces.

rafaelsouzaf committed:
Added Server Auth definitions.

rafaelsouzaf committed:
Updated index with last file definitions.

Commit:
Added more Server definitions.

rafaelsouzaf committed:
Added more Server types

rafaelsouzaf committed:
Creating namespace Lifecycle and moving FailAction to inside.

rafaelsouzaf committed:
Reviewing Route types and getting better. Not finished yet.

rafaelsouzaf committed:
Added Json namespace. Cloned to v16 types.

rafaelsouzaf committed:
Using new Lifecycle namespace.

rafaelsouzaf committed:
Solved some TODO's

rafaelsouzaf committed:
Updated with new types.

rafaelsouzaf committed:
Renamed Lifecycle methods for better understanding.

Commit:
Reviewing Routes type implementation.

Commit:
Added some Lifecycle definitions.

Commit:
Fix. Each lifecycle method must return a value or a promise.

rafaelsouzaf committed:
Reviewed Response declarations.

rafaelsouzaf committed:
Reviewed Request declarations.

rafaelsouzaf committed:
Fix in JsDocs.

rafaelsouzaf committed:
Solved TODO about plugin,

rafaelsouzaf committed:
Fix in JsDocs.

rafaelsouzaf committed:
Added more Server types.

rafaelsouzaf committed:
Reviewed and added sonme Function Overloading to follow best practices: http://definitelytyped.org/guides/best-practices.html

rafaelsouzaf committed:
Reviewed some server types.

rafaelsouzaf committed:
Refactored Json and Util namespaces.

ctaggart committed:
examples replacing any with object

rafaelsouzaf committed:
Merge pull request #3 from ctaggart/hapi-17-object

Replacing any with object

BorntraegerMarc committed:
Added first hapi TS tests

Commit:
Revert "examples replacing any with object"

This reverts commit cb083b4

It was my mistake. The example code is outside of v17 folder.

rafaelsouzaf committed:
Reviewed some server types.

rafaelsouzaf committed:
Added some test files.

rafaelsouzaf committed:
Changed state() method param from string to object.

rafaelsouzaf committed:
Added more tests files.

rafaelsouzaf committed:
RequestExtPointFunction should inherit from Lifecycle.Method

rafaelsouzaf committed:
Added "continue.ts" test file.

gitawego commented:
it seems the param ``value`` for method ``response`` in response-toolkit.d.ts should be ``any`` or at least ``string | object`` imo, depends on [this test](https://github.com/hapijs/hapi/blob/4144f9397cf639e22379699cdfc967d2c84e8509/test/response.js#L220)

rafaelsouzaf committed:
Add overloaded function "response(value: object)" to work with hapijs test:
https://github.com/hapijs/hapi/blob/4144f9397cf639e22379699cdfc967d2c84e8509/test/response.js#L220

Following suggestion of @gitawego

rafaelsouzaf commented:
[Fixed](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/c739273012528ec666398c5366ecdc55861794e9). Thanks @gitawego 

In the same file we have the method `redirect()` with **no-optional (mandatory)** param in the [JSDoc](https://github.com/hapijs/hapi/blob/master/API.md#-hredirecturi), but optional in the [tests](https://github.com/hapijs/hapi/blob/4144f9397cf639e22379699cdfc967d2c84e8509/test/response.js#L846). Do you know what I should consider?

Maybe the best approach to test everything is migrate the official hapijs tests to Typescript.


**EDITED**: The official [hapijs tests](https://github.com/hapijs/hapi/tree/4144f9397cf639e22379699cdfc967d2c84e8509/test) are huge! I can't migrate to TS everything alone. It will take very long time to do this job and maybe it's not necessary. If anybody can help I will appreciate :) For now I will continue with the basic tests using examples from the doc API and based in v16 version.

ctaggart commented:
I'm currently thinking that it is better to use union types as recommended
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

Instead of:
``` ts
     response(value: string): ResponseObject;
     response(value: object): ResponseObject;
```

do this:
``` ts
     response(value: string|object): ResponseObject;
```

rafaelsouzaf commented:
@ctaggart I completely understand and I have read this link and recommendation before.

But, on the other hand, there is another recommendation from [definitelytyped.org](http://definitelytyped.org/guides/best-practices.html). Please see _Function Overloading_ sub-section.

> For functions / constructors use function overloading e.g

```
declare class Foo {
    constructor(foo: number);
    constructor(foo: string);
}
```

So I am confused:-)
What is the best option? I am not sure.

PS: @ctaggart If you can help me reviewing this [PR](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/4) I will appreciate :-)

ctaggart commented:
My plan for review of this API was to try it out from Fable by running it through [ts2fable](https://www.npmjs.com/package/ts2fable) which I've been writing. http://blog.ctaggart.com/search/label/ts2fable

I like `response(value: string|object): ResponseObject;` because that is what is going to work. :-)

rafaelsouzaf committed:
Add object arguments (...args) instead of Function. Is it good idea? TODO created.

rafaelsouzaf committed:
Add new RequestEvents type.

rafaelsouzaf committed:
Add RequestEvents type in index.d.ts

rafaelsouzaf committed:
Separate ServerEvents in a new file. Fixed method overriding from Podium interface. I hope.

rafaelsouzaf committed:
Add events (server and request) test files.

ctaggart commented:
After sleeping on it, I'll modify ts2fable to support what you have already with:
``` ts
     response(value: string): ResponseObject;
     response(value: object): ResponseObject;
```

rafaelsouzaf commented:
LOL great @ctaggart !
Don't kill me if soon we change to `response(value: string | object): ResponseObject` :-)

ctaggart commented:
> Don't kill me if soon we change to response(value: string | object): ResponseObject :-)

👍 I'll make it work either way.

gitawego commented:
@rafaelsouzaf  for ``redirect()`` part, I think we should be based on the test, if their test passes, it means the param ``uri`` should be optional. 

I will try to check as much as I can their test files, and see if there should be any improvement.  but I think your definition is mature enough to send a PR to upstream, once it merged, ppl could contribute, as you said, there are too many tests, you can't do it alone ^^

gitawego commented:
@rafaelsouzaf  for ``response`` part, depends on [official best practice](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html), in ``Use Union Types`` section, it says: ``Don’t write overloads that differ by type in only one argument position``, so I think it's better with ``response(value: string | object): ResponseObject``

rafaelsouzaf committed:
Replacing any with object. Following the idea of @ctaggart (#4)

Review of TS types using 'object' and 'any' mainly.

rafaelsouzaf committed:
Replace Dictionary<T> by Map<T, T> supported in es6

rafaelsouzaf committed:
Add h.redirect() without param to work with hapijs official tests.

rafaelsouzaf commented:
@gitawego commited [redirect()](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/81e58330cca9951cfc9eb9781a6295957af385fb) without params 👍 

Thanks for the tests. I made some, _maybe delicate_, changes today ([1](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/e189cf6b5e09901f6c068fc78960e37fa129ec77), [2](https://github.com/rafaelsouzaf/DefinitelyTyped/pull/1/commits/9fb4ab18b394a9aec15afee86d80588dbba834f1)). If you detect any problem please let me know.

I would like to hear the opinion of my mentor @BorntraegerMarc  :-)

rafaelsouzaf committed:
Add ResponseEvent types.

rafaelsouzaf committed:
Add request.route.auth.access() method

rafaelsouzaf committed:
Add "Function" support param in server.events.emit()

rafaelsouzaf committed:
Optimize imports.

rafaelsouzaf committed:
Going back with Dictionary. It didn't work as expected :-/

rafaelsouzaf committed:
Add more tests files.

rafaelsouzaf committed:
Add more test files.

rafaelsouzaf committed:
Fix decorate() param. It must support the "toolkit" option.

rafaelsouzaf committed:
Fix mime property. It must be "any" like v16.

rafaelsouzaf committed:
Many fix discovered with the tests.

rafaelsouzaf committed:
Many fix discovered with the tests cases.

rafaelsouzaf committed:
Add more new test files, mainly Server test files.

rafaelsouzaf committed:
Add new type ServerCache in index.

rafaelsouzaf committed:
More fix based in test cases.

rafaelsouzaf committed:
Add more tests.

rafaelsouzaf commented:
Ok. A lot of tests were built based on v17 API code samples. And some v16 tests have been adapted to the new version.
It's pending the `response(value: string | object)` change.

BorntraegerMarc commented on types/hapi/v17/test/request/event-types.ts:
Why is this not an arrow function? `(request: Request, h: ResponseToolkit) => {`

BorntraegerMarc commented on types/hapi/v17/test/request/event-types.ts:
In general there are quite a bit of "normal" functions. In my opinion the tests should follow one standard. Ether `function` or `() =>`    

I  personally like the newer arrow function standard better 😄

BorntraegerMarc commented on types/hapi/v17/test/response/response-events.ts:
Comments probably can be deleted?

BorntraegerMarc commented on types/hapi/v17/test/route/config.ts:
Can this comment be deleted?

BorntraegerMarc commented on types/hapi/v17/test/server/server-cache-provision.ts:
What exactly is not working?

BorntraegerMarc commented on types/hapi/v17/test/server/server-decorations.ts:
Not sure where the problem is here. Could you specify?

BorntraegerMarc commented on types/hapi/v17/test/server/server-info.ts:
Do the `assert` checks actually get executed? I thought the tests were mainly to see if it compiles without errors? Not to actually make unit tests. Unit tests seem to be a bit of an overkill for me because the whole code is already unit tested in the "regular" hapijs project. So my assumption is -> if it compiles correctly to JS -> then the "normal" unit tests will have tested the code correctly anyway. That is why `assert`'s are not mandatory :)

BorntraegerMarc commented on types/hapi/v17/definitions/server/server-options.d.ts:
Maybe one test where all the options of this interface is used, would be nice?

BorntraegerMarc reviewed (approved):
So first of all let me say: Great work! I really appreciate the work you went through @rafaelsouzaf 🎉 Everyone is going to benefit from these changes :)

I agree with @gitawego -> I think the PR is in a state where it is good to be merged into the official DefinitelyTyped repository. From there we will see again some more reviews from early adopters and maybe new PR's will come up to address these reviews.

From my side: we are good to go to merge the PR 😄 

PS: Personally I like the style `response(value: string | object)` better :)

rafaelsouzaf committed:
Delete comments.

rafaelsouzaf committed:
Remove asserts. It's not necessary here.

rafaelsouzaf committed:
Replace all "function" with arrow function "() =>".

rafaelsouzaf committed:
Add "server-options.ts" test file containing all possible options.

rafaelsouzaf committed:
Add type object to compression property. It is not only boolean.

rafaelsouzaf committed:
Add more options in "server-options.ts" test file.

rafaelsouzaf commented on types/hapi/v17/definitions/server/server-options.d.ts:
@BorntraegerMarc very good idea. I created this [test file](https://github.com/rafaelsouzaf/DefinitelyTyped/blob/6a92487939470e96f6f3219963e66962a97ff6ac/types/hapi/v17/test/server/server-options.ts).

rafaelsouzaf commented on types/hapi/v17/test/request/event-types.ts:
Commit 1a5f4f5

rafaelsouzaf commented on types/hapi/v17/test/response/response-events.ts:
Commit 6a11db6

rafaelsouzaf commented on types/hapi/v17/test/server/server-cache-provision.ts:
All the examples I am compiling (obviously) AND executing with NodeJS. TS compiler is good but some problems I can see only when I execute it. In this particular case, there is an exception in execution time. So I am not sure if the problem is my definitions, is the example or the hapijs framework.

I think the problem is with the example code. But I am not sure. I would like to confirm it with someone.

```AssertionError [ERR_ASSERTION]: Unknown cache countries
    at module.exports.internals.Core._cachePolicy (D:\Workspace\DefinitelyTyped-Rafael\node_modules\hapi\lib\core.js:520:14)
    at internals.Server.policy [as cache] (D:\Workspace\DefinitelyTyped-Rafael\node_modules\hapi\lib\server.js:513:27)
    at Object.<anonymous> (D:\Workspace\DefinitelyTyped-Rafael\types\hapi\v17\dist\server\server-cache-provision.js:10:20)
    at Module._compile (module.js:641:30)
    at Object.Module._extensions..js (module.js:652:10)
    at Module.load (module.js:560:32)
    at tryModuleLoad (module.js:503:12)
    at Function.Module._load (module.js:495:3)
    at Function.Module.runMain (module.js:682:10)
    at startup (bootstrap_node.js:191:16)
[nodemon] app crashed - waiting for file changes before starting...```

rafaelsouzaf commented on types/hapi/v17/test/server/server-decorations.ts:
My mistake. Fixed.

rafaelsouzaf commented on types/hapi/v17/test/server/server-info.ts:
@BorntraegerMarc Yes, you are right. The most important is compiling. I am executing as well but the assert's is too much. 3d6b184

rafaelsouzaf commented on types/hapi/v17/test/route/config.ts:
Commit 6a11db6

rafaelsouzaf commented:
> So first of all let me say: Great work! I really appreciate the work you went through @rafaelsouzaf 🎉 Everyone is going to benefit from these changes :)

Thanks @BorntraegerMarc . It was a very nice experience and I learned a lot. Thank you!

> I agree with @gitawego -> I think the PR is in a state where it is good to be merged into the official DefinitelyTyped repository.

Perfect. I will come back with the "method overriding" and prepare the PR to upstream.
👍 👍

BorntraegerMarc commented on types/hapi/v17/test/server/server-cache-provision.ts:
Did you copy it from here https://github.com/hapijs/hapi/blob/master/API.md#-await-servercacheprovisionoptions ? Because one difference I see is that the example uses `await`, which we don't in our test, here...

Commit:
Following best practices of Function Overloading: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

Commit:
Add RouteOptions test file with complete available options.
